PrincessBrideTrivia/PrincessBrideTrivia/Program.cs

```diff
@@ -1,27 +1,79 @@
+    {
+        Accept,
+        Reject,
+        Invalid
+    }
+
+    private const int QuestionBlockSize = 5;
+
+    private const int AnswerOptionsGiven = QuestionBlockSize - 2;
+    // since we know that each question block contains both a question text and a correct answer, we can get the number
+    // of questions expected by subtracting 2 from the question block size
+
     public static void Main(string[] args)
     {
+        const int maxAttempts = 3;
+
         string filePath = GetFilePath();
         Question[] questions = LoadQuestions(filePath);
-
         int numberCorrect = 0;
-        for (int i = 0; i < questions.Length; i++)
+        int[] attemptScores = new int[maxAttempts];
+        for (int attemptIndex = 0; attemptIndex < maxAttempts; attemptIndex++)
         {
-            bool result = AskQuestion(questions[i]);
-            if (result)
+            foreach (Question question in questions)
+            {
+                bool result = AskQuestion(question);
+                if (result)
+                {
+                    numberCorrect++;
+                }
+            }
+            attemptScores[attemptIndex] = numberCorrect;
+            Console.WriteLine($"You got {GetPercentCorrect(numberCorrect, questions.Length)} correct");
+            numberCorrect = 0;
+            if (attemptIndex != 2)
             {
-                numberCorrect++;
+                Console.WriteLine("Do you want to make another attempt?");
+                Console.WriteLine("1 : Yes");
+                Console.WriteLine("2 : No");
+                ResponseValidation responseValidation = ResponseValidation.Invalid;
+                while (responseValidation == ResponseValidation.Invalid) 
+                { 
+                    responseValidation = AcceptRetryQuiz(Console.ReadLine());
+                }
+                if (responseValidation == ResponseValidation.Reject)
+                {
+                    break;
+                }
+                else if(responseValidation == ResponseValidation.Accept)
+                {
+                    Random random = new Random();
+                    questions = questions.OrderBy(x => random.Next()).ToArray();
+                    foreach (Question question in questions)
+                    {
+                        question.RandomizeAnswerOrder(random.Next());
+                    }
+                }
             }
         }
-        Console.WriteLine("You got " + GetPercentCorrect(numberCorrect, questions.Length) + " correct");
+        Console.WriteLine($"Your final score is {GetPercentCorrect(attemptScores.Max(), questions.Length)}");
     }
 
     public static string GetPercentCorrect(int numberCorrectAnswers, int numberOfQuestions)
     {
-        return (numberCorrectAnswers / numberOfQuestions * 100) + "%";
+        if (numberOfQuestions == 0)
+            return "N/A";
+
+        double precent = (double)numberCorrectAnswers / numberOfQuestions * 100;
```

Comment:
Just pointing out a small typo here and on the following line. It should be 'percent'.